src/mscorlib/shared/System/IO/UnmanagedMemoryStream.cs

```diff
@@ -583,17 +604,40 @@ public override void Write(byte[] buffer, int offset, int count)
                 throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_NeedNonNegNum);
             if (buffer.Length - offset < count)
                 throw new ArgumentException(SR.Argument_InvalidOffLen);
-            Contract.EndContractBlock();  // Keep contract validation in sync with WriteAsync(..)
 
-            if (!_isOpen) throw Error.GetStreamIsClosed();
-            if (!CanWrite) throw Error.GetWriteNotSupported();
+            WriteCore(new Span<byte>(buffer, offset, count));
+        }
+
+        public override void Write(ReadOnlySpan<byte> source)
+        {
+            if (GetType() == typeof(UnmanagedMemoryStream))
+            {
+                WriteCore(source);
+            }
+            else
+            {
+                // UnmanagedMemoryStream is not sealed, and a derived type may have overridden Write(byte[], int, int) prior
```

Comment:
> Is your thinking that we implement more efficient version of these overloads for all CoreLib/CoreFX streams, or just for the ones that matter (e.g. public Streams)?

Starting with the ones that matter, and moving on to the rest :)  The best solution I've come up with for addressing the derived cases is exposing the actual functionality in a protected method, which would be called both from this Read method and could be called from a derived type if it decided to delegate to the base.  But as that involves exposing even more public surface area, I think we should hold off on it unless it's very necessary.  If the streams in question are in the same assembly, though, we can and should do so with an internal method, if there's a reasonable possibility that the stream will be exposed to code that'll consume the new methods.  I see now that it looks like UnmanagedMemoryStreamWrapper and PinnedBufferMemoryStream fall into this camp in CoreLib, so I'll update those accordingly.